source/blog/2024-11-05-hanami-220.html.markdown

```diff
@@ -0,0 +1,244 @@
+          post_repo.latest(page:)
+        end
+      end
+    end
+  end
+end
+```
+
+Repo methods return structs: plain old value objects, just data, with no live connection back to the database. This means you can be confident in passing them all around your app, knowing things like accidental n+1 queries are a thing of the past.
+
+You can customize these structs by creating your own matching classes. Make one for your posts with `hanami generate struct post`. Inside these classes, you can access any of the attributes selected in your repo’s corresponding database query.
+
+```ruby
+module MyApp
+  module Structs
+    class Post < MyApp::DB::Struct
+      def excited_title
+        "#{title}!"
+      end
+    end
+  end
+end
+```
+
+With relations, repos, structs and more, you now have a home for every piece of your data logic, and the foundation for a database layer that can evolve to meet even the most demanding needs.
+
+Hanami’s database capabilities come from [ROM](https://rom-rb.org) (Ruby Object Mapper), a mature and flexible data persistence toolkit for Ruby. With Hanami 2.2, we are proud to provide the world’s very best ROM experience, one that feels thorouhgly at home within Hanami apps. We make it easy for you to get started and enjoy the benefits of a dedicated persistence layer, while making sure you can still use every ROM feature without ever having to eject yourself from our integration of ROM with Hanami.
+
+## Your DB is our command
+
+You can manage the full lifecycle of your database thanks to this complete set of new CLI commands:
+
+- `hanami db create`
+- `hanami db drop`
+- `hanami db migrate`
+- `hanami db prepare`
+- `hanami db seed`
+- `hanami db structure dump`
+- `hanami db structure load`
+- `hanami db version`
+
+## Slice it your way
+
+It wouldn’t be a new Hanami feature if it didn’t come with first-class support for [slices](https://guides.hanamirb.org/v2.2/app/slices/), our built-in tool for modularizing your apps.
+
+With the new database layer, you can choose your own mix of one or more databases across slices. You can choose to:
+
+- Share a single database, but have each slice provide its own relations, so they can choose exactly how much of the database to expose. This is our default.
+- Use a dedicated database for certain slices. This is as easy as a `SLICE_NAME__DATABASE_URL` env var!
+- Connect to any number of _additional_ databases within your app or slice. This is as easy as creating a `SLICE_NAME__DATABASE_URL__GATEWAY_NAME` env var!
+- Adopt a simpler, blended development experience by sharing a single database and set of relations across all slices.
+- Or any combination of the above!
+
+## Smooth operations
+
+Along with Hanami 2.2, we’re excited to debut the brand new [dry-operation](https://dry-rb.org/gems/dry-operation) gem. With dry-operation, you have a streamlined tool for organizing your business logic into flexible, composable objects made from flows of internal steps.
```

Comment:
Should we add something here about how dry-operation is a first-class part of the Hanami stack, but an optional one? People are welcome to use other libraries/their own

Reply:
Added a paragraph on this. Let me know what you think.